2_process/src/raster_to_catchment_polygons.R

```diff
@@ -0,0 +1,89 @@
+}
+
+raster_to_catchment_polygons <- function(polygon_sf, raster,
+                                         categorical_raster = NULL,
+                                         raster_summary_fun = NULL,
+                                         new_cols_prefix = NULL){
+
+  #' @description Function extracts raster pixel per catchment polygon and calculated proportion of raster class per total
+  #'
+  #' @param polygon_sf sf object of polygon(s) in aoi
+  #' @param raster path to '.tif' file or raster object of SpatRaster object  
+  #' @param categorical_raster logical. If categorical raster, TRUE
+  #' @param raster_summary_fun for continuous raster, function to summarize the data by geometry. kept NULL if categorical_raster = TRUE
+  #' @param categorical_raster_prefix for categorical raster, prefix added to colnames to label raster classes/values
+  #' @value A data frame the same columns as the polygon_sf spatial dataframe with additional columns of the Land Cover classes
+  #' @example land cover raster ex: raster_to_catchment_polygons(polygon_sf = p1_catchments_sf, raster = p1_backcasted_LC[1], categorical_raster = TRUE, new_cols_prefix = 'lc_)
+  #' @example cont. raster ex: raster_to_catchment_polygons(polygon_sf = p1_catchments_sf, raster = p1_rod_salt, categorical_raster = FALSE, raster_summary_fun = sum, new_cols_prefix = 'cont_raster')
+  
+  # read in 
+  raster <- rast(raster)
+  vector_sf <- polygon_sf 
+  
+  ## check vector geometries 
+  if(any(!st_is_valid(vector_sf))){
+    vector_sf <- st_make_valid(vector_sf)
+    message('shp geometries fixed')
+    } 
+  
+  ## match crs
+  if(!st_crs(raster) == st_crs(vector_sf)){
+    message('crs are different. Transforming ...')
+    vector_sf <- st_transform(vector_sf, crs = st_crs(raster))
+    if(st_crs(raster) == st_crs(vector_sf)){
+    message('crs now aligned')}
+  }else if(st_crs(raster) == st_crs(vector_sf)){
+    message('crs are already aligned')
+  }
+  
+  ## convert vector sf object to spatvector compatible to with raster processing with terra
+  vector <- vector_sf %>% vect()
+
+  ## crop raster to catchment_sf 
+  raster_crop <- terra::crop(raster, vector)
+  
+  if(categorical_raster == TRUE){
+  ## Extract rasters pixels in each polygon
+  start_time <- Sys.time()
+  raster_per_polygon <- terra::extract(raster_crop, vector, list = T) %>% 
+    #lapply(table) # to get frequency of each categorical value
+    lapply(FUN = function(x) {table(x)/sum(table(x))})
+  end_time <- Sys.time()
+  message('raster extract processing time:', end_time - start_time)
+  
+  ## handling unequal length of classes
+  final_raster_table <- rbind.fill.list(raster_per_polygon) %>%
+    setNames(paste0(new_cols_prefix, names(.))) %>% 
+    mutate(ID = row_number())
+  
+  }
+  
+  else{
+    message('extracting cont raster')
+    raster_per_polygon <- terra::extract(raster_crop, vector, fun = raster_summary_fun)
+    print(head(raster_per_polygon))
```

Comment:
Was this `print` statement mostly for exploratory purposes? Is it OK to remove it to clean up the pipeline output?